src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs

```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
             await client.ApiClient.RemoveAllReactionsAsync(msg.Channel.Id, msg.Id, options);
         }
 
-        public static async Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
+        public static IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
             Action<GetReactionUsersParams> func, BaseDiscordClient client, RequestOptions options)
         {
-            var args = new GetReactionUsersParams();
-            func(args);
-            string emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
-            return (await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false)).Select(u => RestUser.Create(client, u)).ToImmutableArray();
+            Preconditions.NotNull(emote, nameof(emote));
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
+
+            var arguments = new GetReactionUsersParams();
+            func(arguments);
+
+            return new PagedAsyncEnumerable<IUser>(
+                DiscordConfig.MaxUserReactionsPerBatch,
+                async (info, ct) =>
+                {
+                    var args = new GetReactionUsersParams();
+                    func(args);
+                    args.Limit = info.PageSize;
+
+                    if (info.Position != null)
+                        args.AfterUserId = info.Position.Value;
+
+                    var models = await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false);
+                    var builder = ImmutableArray.CreateBuilder<IUser>();
```

Comment:
IMO this should be replaced with models.Select and you should call ToImmutableArray on it